File: backend-spring/src/main/java/com/ivyts/backend/web/auth/AuthController.java
package com.ivyts.backend.web.auth;

import com.ivyts.backend.common.api.ApiSuccessResponse;
import com.ivyts.backend.security.RequestAuthService;
import com.ivyts.backend.service.AuthService;
import com.ivyts.backend.service.GoogleOAuthService;
import com.ivyts.backend.web.auth.dto.ChangePasswordRequest;
import com.ivyts.backend.web.auth.dto.ConfirmEmailChangeRequest;
import com.ivyts.backend.web.auth.dto.ConfirmPhoneChangeRequest;
import com.ivyts.backend.web.auth.dto.GoogleAuthRequest;
import com.ivyts.backend.web.auth.dto.LoginRequest;
import com.ivyts.backend.web.auth.dto.RefreshTokenRequest;
import com.ivyts.backend.web.auth.dto.RegisterRequest;
import com.ivyts.backend.web.auth.dto.RequestEmailChangeRequest;
import com.ivyts.backend.web.auth.dto.RequestPhoneChangeRequest;
import com.ivyts.backend.web.auth.dto.UpdateProfileRequest;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@Validated
@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final AuthService authService;
    private final GoogleOAuthService googleOAuthService;
    private final RequestAuthService requestAuthService;

    public AuthController(AuthService authService, GoogleOAuthService googleOAuthService, RequestAuthService requestAuthService) {
        this.authService = authService;
        this.googleOAuthService = googleOAuthService;
        this.requestAuthService = requestAuthService;
    }

    @PostMapping("/register")
    @ResponseStatus(HttpStatus.CREATED)
    public ApiSuccessResponse<?> register(@Valid @RequestBody RegisterRequest body) {
        return ApiSuccessResponse.of("Register successful", authService.register(body));
    }

    @PostMapping("/login")
    public ApiSuccessResponse<?> login(@Valid @RequestBody LoginRequest body) {
        return ApiSuccessResponse.of("Login successful", authService.login(body));
    }

    @PostMapping("/google")
    public ApiSuccessResponse<?> googleAuth(@Valid @RequestBody GoogleAuthRequest body) {
        return ApiSuccessResponse.of("Google login successful", googleOAuthService.authenticate(body));
    }

    @PostMapping("/logout")
    public ApiSuccessResponse<?> logout(HttpServletRequest request) {
        authService.logout(requestAuthService.requireUser(request));
        return ApiSuccessResponse.of("Logout successful", Map.of());
    }

    @GetMapping("/me")
    public ApiSuccessResponse<?> me(HttpServletRequest request) {
        return ApiSuccessResponse.of("Current user fetched successfully", authService.me(requestAuthService.requireUser(request)));
    }

    @PostMapping("/refresh-token")
    public ApiSuccessResponse<?> refreshToken(@Valid @RequestBody RefreshTokenRequest body) {
        return ApiSuccessResponse.of("Refresh token successful", authService.refreshToken(body));
    }

    @PatchMapping("/me/profile")
    public ApiSuccessResponse<?> updateProfile(HttpServletRequest request, @Valid @RequestBody UpdateProfileRequest body) {
        return ApiSuccessResponse.of("Profile updated successfully", authService.updateProfile(requestAuthService.requireUser(request), body));
    }

    @PostMapping("/me/password")
    public ApiSuccessResponse<?> changePassword(HttpServletRequest request, @Valid @RequestBody ChangePasswordRequest body) {
        authService.changePassword(requestAuthService.requireUser(request), body);
        return ApiSuccessResponse.of("Password updated successfully", Map.of());
    }

    @PostMapping("/me/email-change/request")
    public ApiSuccessResponse<?> requestEmailChange(HttpServletRequest request, @Valid @RequestBody RequestEmailChangeRequest body) {
        return ApiSuccessResponse.of("Email verification code sent successfully", authService.requestEmailChange(requestAuthService.requireUser(request), body));
    }

    @PostMapping("/me/email-change/confirm")
    public ApiSuccessResponse<?> confirmEmailChange(HttpServletRequest request, @Valid @RequestBody ConfirmEmailChangeRequest body) {
        return ApiSuccessResponse.of("Email updated successfully", authService.confirmEmailChange(requestAuthService.requireUser(request), body));
    }

    @PostMapping("/me/phone-change/request")
    public ApiSuccessResponse<?> requestPhoneChange(HttpServletRequest request, @Valid @RequestBody RequestPhoneChangeRequest body) {
        return ApiSuccessResponse.of("Phone OTP sent successfully", authService.requestPhoneChange(requestAuthService.requireUser(request), body));
    }

    @PostMapping("/me/phone-change/confirm")
    public ApiSuccessResponse<?> confirmPhoneChange(HttpServletRequest request, @Valid @RequestBody ConfirmPhoneChangeRequest body) {
        return ApiSuccessResponse.of("Phone updated successfully", authService.confirmPhoneChange(requestAuthService.requireUser(request), body));
    }
}

File: archive/backend-node/src/controllers/auth.controller.ts
import { Request, Response } from 'express';
import { HTTP_STATUS } from '../constants/http-status.js';
import {
  changeCurrentUserPassword,
  confirmCurrentUserEmailChange,
  confirmCurrentUserPhoneChange,
  getCurrentUser,
  loginUser,
  logoutUser,
  refreshUserToken,
  registerUser,
  requestCurrentUserEmailChange,
  requestCurrentUserPhoneChange,
  updateCurrentUserProfile,
} from '../services/auth.service.js';
import { emitActionFailure } from '../services/notification-events.service.js';
import { sendSuccess } from '../utils/api-response.js';

export async function register(req: Request, res: Response): Promise<Response> {
  const data = await registerUser(req.body);
  return sendSuccess(res, HTTP_STATUS.CREATED, 'Register successful', data);
}

export async function login(req: Request, res: Response): Promise<Response> {
  const data = await loginUser(req.body);
  return sendSuccess(res, HTTP_STATUS.OK, 'Login successful', data);
}

export async function logout(req: Request, res: Response): Promise<Response> {
  await logoutUser(req.body);
  return sendSuccess(res, HTTP_STATUS.OK, 'Logout successful', {});
}

export async function me(req: Request, res: Response): Promise<Response> {
  const data = await getCurrentUser(req.user!.userId);
  return sendSuccess(res, HTTP_STATUS.OK, 'Current user fetched successfully', data);
}

export async function refreshToken(req: Request, res: Response): Promise<Response> {
  const data = await refreshUserToken(req.body);
  return sendSuccess(res, HTTP_STATUS.OK, 'Refresh token successful', data);
}

export async function updateProfile(req: Request, res: Response): Promise<Response> {
  try {
    const data = await updateCurrentUserProfile(req.user!.userId, req.body);
    return sendSuccess(res, HTTP_STATUS.OK, 'Profile updated successfully', data);
  } catch (error) {
    emitActionFailure(req.user!, 'Cap nhat thong tin that bai', 'Khong the cap nhat thong tin ca nhan luc nay.', 'profile');
    throw error;
  }
}

export async function changePassword(req: Request, res: Response): Promise<Response> {
  try {
    await changeCurrentUserPassword(req.user!.userId, req.body);
    return sendSuccess(res, HTTP_STATUS.OK, 'Password updated successfully', {});
  } catch (error) {
    emitActionFailure(req.user!, 'Doi mat khau that bai', 'Khong the cap nhat mat khau voi thong tin vua nhap.', 'profile');
    throw error;
  }
}

export async function requestEmailChange(req: Request, res: Response): Promise<Response> {
  const data = await requestCurrentUserEmailChange(req.user!.userId, req.body);
  return sendSuccess(res, HTTP_STATUS.OK, 'Email verification code sent successfully', data);
}

export async function confirmEmailChange(req: Request, res: Response): Promise<Response> {
  try {
    const data = await confirmCurrentUserEmailChange(req.user!.userId, req.body);
    return sendSuccess(res, HTTP_STATUS.OK, 'Email updated successfully', data);
  } catch (error) {
    emitActionFailure(req.user!, 'Cap nhat email that bai', 'Khong the xac nhan thay doi email.', 'profile');
    throw error;
  }
}

export async function requestPhoneChange(req: Request, res: Response): Promise<Response> {
  const data = await requestCurrentUserPhoneChange(req.user!.userId, req.body);
  return sendSuccess(res, HTTP_STATUS.OK, 'Phone OTP sent successfully', data);
}

export async function confirmPhoneChange(req: Request, res: Response): Promise<Response> {
  try {
    const data = await confirmCurrentUserPhoneChange(req.user!.userId, req.body);
    return sendSuccess(res, HTTP_STATUS.OK, 'Phone updated successfully', data);
  } catch (error) {
    emitActionFailure(req.user!, 'Cap nhat so dien thoai that bai', 'Khong the xac nhan thay doi so dien thoai.', 'profile');
    throw error;
  }
}
